packages/periphery/contracts/oracles/UniswapV2Oracle.sol

```diff
@@ -0,0 +1,182 @@
+  error WrongGranularity();
+  error WrongWindow();
+  error PairNotFound();
+  error MissingHistoricalObservation();
+  error UnexpectedTimeElapsed();
+  error PairAlreadyExists();
+
+  struct TokenPair {
+    address token0;
+    address token1;
+  }
+
+  struct Observation {
+    uint timestamp;
+    uint price0Cumulative;
+  }
+
+  struct PairOracleOptions {
+    uint16 secondsAgo;
+    uint16 secondsAgoLiquidation;
+  }
+
+  address public immutable factory;
+  // the desired amount of time over which the moving average should be computed, e.g. 24 hours
+  uint public immutable windowSize;
+  // the number of observations stored for each pair, i.e. how many price observations are stored for the window.
+  // as granularity increases from 1, more frequent updates are needed, but moving averages become more precise.
+  // averages are computed over intervals with sizes in the range:
+  //   [windowSize - (windowSize / granularity) * 2, windowSize]
+  // e.g. if the window size is 24 hours, and the granularity is 24, the oracle will return the average price for
+  //   the period:
+  //   [now - [22 hours, 24 hours], now]
+  uint8 public immutable granularity;
+  // this is redundant with granularity and windowSize, but stored for gas savings & informational purposes.
+  uint public immutable periodSize;
+
+  // mapping from pair address to a list of price observations of that pair
+  mapping(address => Observation[]) public pairObservations;
+  address[] public pairs;
+  mapping(address => PairOracleOptions) public pairOptions;
+
+  constructor(address factory_, uint windowSize_, uint8 granularity_) {
+    if (factory_ == address(0)) revert WrongValue();
+    if (granularity_ < 2) revert WrongGranularity();
+    if ((periodSize = windowSize_ / granularity_) * granularity_ != windowSize_) revert WrongWindow();
+
+    factory = factory_;
+    windowSize = windowSize_;
+    granularity = granularity_;
+  }
+
+  ///@notice Returns the index of the observation corresponding to the given timestamp
+  function observationIndexOf(uint timestamp) public view returns (uint8 index) {
+    uint epochPeriod = timestamp / periodSize;
+    return uint8(epochPeriod % granularity);
+  }
+
+  ///@notice update the cumulative price for the observation at the current timestamp. each observation is updated at most once per epoch period
+  function update(address pair) public {
+    // get the observation for the current period
+    uint8 observationIndex = observationIndexOf(block.timestamp);
+    Observation storage observation = pairObservations[pair][observationIndex];
+
+    // we only want to commit updates once per period (i.e. windowSize / granularity)
+    uint timeElapsed = block.timestamp - observation.timestamp;
+    if (timeElapsed > periodSize) {
+      (uint price0Cumulative, , ) = UniswapV2OracleLibrary.currentCumulativePrices(pair);
+      observation.timestamp = block.timestamp;
+      observation.price0Cumulative = price0Cumulative;
```

Comment:
so basically you'll need to implement here something like `observation.priceCumulative = token0IsBaseToken ? price0Cumulative : price1Cumulative`